src/Shared/PercentEncodingHelper.cs

```diff
@@ -0,0 +1,146 @@
+    private const char ListSplitter = ',';
+
+    internal static bool TryExtractBaggage(
+        string[] baggageCollection,
+#if NET
+        [NotNullWhen(true)]
+#endif
+        out Dictionary<string, string>? baggage)
+    {
+        Dictionary<string, string>? baggageDictionary = null;
+        int baggageLength = -1; // Start with -1 to account for no leading comma on first item
+
+        foreach (var baggageList in baggageCollection.Where(h => !string.IsNullOrEmpty(h)))
+        {
+            foreach (string keyValuePair in baggageList.Split(ListSplitter))
+            {
+                baggageLength += keyValuePair.Length + 1; // pair length + comma
+                if (ExceedsMaxBaggageLimits(baggageLength, baggageDictionary?.Count))
+                {
+                    baggage = baggageDictionary;
+                    return baggageDictionary != null;
+                }
+#if NET
+                var indexOfFirstEquals = keyValuePair.IndexOf(KeyValueSplitter, StringComparison.Ordinal);
+#else
+                var indexOfFirstEquals = keyValuePair.IndexOf(KeyValueSplitter);
+#endif
+                if (indexOfFirstEquals < 0)
+                {
+                    continue;
+                }
+
+                var splitKeyValue = keyValuePair.Split([KeyValueSplitter], 2);
+                var key = splitKeyValue[0].Trim();
+                var value = splitKeyValue[1].Trim();
+
+                if (!IsValidKeyValuePair(key, value))
+                {
+                    continue;
+                }
+
+                var decodedValue = PercentDecodeBaggage(value);
+
+                baggageDictionary ??= [];
+                baggageDictionary[key] = decodedValue;
+            }
+        }
+
+        baggage = baggageDictionary;
+        return baggageDictionary != null;
+    }
+
+    /// <summary>
+    /// As per the specification, only the value is percent-encoded.
+    /// "Uri.EscapeDataString" encodes code points which are not required to be percent-encoded.
+    /// </summary>
+    /// <param name="key"> The baggage key. </param>
+    /// <param name="value"> The baggage value. </param>
+    /// <returns> The percent-encoded baggage item. </returns>
+    internal static string PercentEncodeBaggage(string key, string value) => $"{key.Trim()}={Uri.EscapeDataString(value.Trim())}";
+
+    private static string PercentDecodeBaggage(string baggageEncoded)
+    {
+        var bytes = new List<byte>();
+        for (int i = 0; i < baggageEncoded.Length; i++)
+        {
+            if (baggageEncoded[i] == '%' && i + 2 < baggageEncoded.Length && IsHex(baggageEncoded[i + 1]) && IsHex(baggageEncoded[i + 2]))
+            {
+                var hex = baggageEncoded.AsSpan(i + 1, 2);
+#if NET
+                bytes.Add(byte.Parse(hex, System.Globalization.NumberStyles.HexNumber, System.Globalization.CultureInfo.InvariantCulture));
+#else
+                bytes.Add(Convert.ToByte(hex.ToString(), 16));
```

Comment:
It will be a lot of string allocations to generate a string for each hex number in the hot path.